pkg/machinery/config/encoder/markdown.go

```diff
@@ -60,86 +28,119 @@ type FileDoc struct {
 		Parse(markdownTemplate))
 
-	buf := bytes.Buffer{}
+	var buf bytes.Buffer
 
-	if err := fd.t.Execute(&buf, fd); err != nil {
+	if err := t.Execute(&buf, struct {
+		Root  *Doc
+		Types map[string]*Doc
+	}{
+		Root:  root,
+		Types: fd.Types,
+	}); err != nil {
 		return nil, err
 	}
 
 	return buf.Bytes(), nil
 }
 
 // Write dumps documentation string to folder.
-func (fd *FileDoc) Write(path, frontmatter string) error {
-	data, err := fd.Encode()
-	if err != nil {
-		return err
-	}
-
-	if stat, e := os.Stat(path); !os.IsNotExist(e) {
+//
+//nolint:gocyclo
+func (fd *FileDoc) Write(path string, frontmatter func(title, description string) string) error {
+	if stat, err := os.Stat(path); !os.IsNotExist(err) {
 		if !stat.IsDir() {
 			return fmt.Errorf("destination path should be a directory")
 		}
 	} else {
-		if e := os.MkdirAll(path, 0o777); e != nil {
-			return e
+		if err := os.MkdirAll(path, 0o777); err != nil {
+			return err
 		}
 	}
 
-	f, err := os.Create(filepath.Join(path, fmt.Sprintf("%s.%s", strings.ToLower(fd.Name), "md")))
-	if err != nil {
+	// generate _index.md
+	if err := os.WriteFile(filepath.Join(path, "_index.md"), []byte(frontmatter(fd.Name, fd.Description)), 0o666); err != nil {
 		return err
 	}
 
-	if _, err := f.WriteString(frontmatter); err != nil {
-		return err
-	}
+	// find map of all types
+	fd.Types = map[string]*Doc{}
 
-	if _, err := f.Write(data); err != nil {
-		return err
+	for _, t := range fd.Structs {
+		if t.Type == "" || strings.ToLower(t.Type) == t.Type {
+			continue
+		}
+
+		fd.Types[t.Type] = t
 	}
 
-	return nil
-}
+	// find root nodes
+	var roots []*Doc
 
-func (fd *FileDoc) encodeType(t string) string {
-	re := regexp.MustCompile(`\w+`)
+	for _, t := range fd.Structs {
+		if len(t.AppearsIn) == 0 {
+			roots = append(roots, t)
+		}
+	}
+
+	for _, root := range roots {
+		contents, err := fd.Encode(root, frontmatter)
+		if err != nil {
+			return err
+		}
 
-	for _, s := range re.FindAllString(t, -1) {
-		if anchor, ok := fd.Anchors[s]; ok {
-			t = strings.ReplaceAll(t, s, formatLink(s, "#"+anchor))
+		if err := os.WriteFile(filepath.Join(path, fmt.Sprintf("%s.%s", strings.ToLower(root.Type), "md")), contents, 0o666); err != nil {
+			return err
 		}
 	}
 
-	return t
+	return nil
 }
 
-func encodeYaml(in interface{}, name string) string {
-	if name != "" {
-		in = map[string]interface{}{
-			name: in,
+//nolint:gocyclo
+func encodeYaml(in any, path string) string {
+	if path != "" {
+		parts := strings.Split(path, ".")
+
+		parts = parts[1:] // strip first segment, it's root element
+
+		// if the last element is ""/"-", it means we're at the root of the slice, so we don't need to wrap it once again
+		if len(parts) > 0 && (parts[len(parts)-1] == "" || parts[len(parts)-1] == "-") {
+			parts = parts[:len(parts)-1]
+		}
+
+		slices.Reverse(parts)
+
+		for _, part := range parts {
+			switch part {
+			case "":
+				in = []any{in}
+			case "-":
+				in = map[string]any{
+					"example.com": in,
```

Comment:
How is this rendered? Is it ignored?  Should we use some constant here?

Reply:
an example is here: https://deploy-preview-8033--talos-dev.netlify.app/v1.6/reference/configuration/v1alpha1/config/#Config.machine.registries.config.-.tls

the only time we have it in v1alpha1 is the registry config, where the key is the hostname of some sort.

when we render inner example, to make it an example rooted at the full config, we use this fake key